pkg/planner/core/plan_cache_instance.go

```diff
@@ -0,0 +1,220 @@
+// Package core Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"go.uber.org/atomic"
+)
+
+// InstancePlanCache represents the instance/node level plan cache.
+type InstancePlanCache interface {
+	// Get gets the cached value from the cache according to key and opts.
+	Get(sctx sessionctx.Context, key string, opts any) (value any, ok bool)
+	// Put puts the key and value into the cache.
+	Put(sctx sessionctx.Context, key string, value, opts any) (succ bool)
+	// Evict evicts some cached values.
+	Evict(sctx sessionctx.Context) (evicted bool)
+	// MemUsage returns the total memory usage of this plan cache.
+	MemUsage(sctx sessionctx.Context) int64
+}
+
+// NewInstancePlanCache creates a new instance level plan cache.
+func NewInstancePlanCache(softMemLimit, hardMemLimit int64) InstancePlanCache {
+	planCache := new(instancePlanCache)
+	planCache.softMemLimit.Store(softMemLimit)
+	planCache.hardMemLimit.Store(hardMemLimit)
+	return planCache
+}
+
+type instancePCNode struct {
+	value    *PlanCacheValue
+	lastUsed atomic.Time
+	next     atomic.Pointer[instancePCNode]
+}
+
+// instancePlanCache is a lock-free implementation of InstancePlanCache interface.
+// [key1] --> [headNode1] --> [node1] --> [node2] --> [node3]
+// [key2] --> [headNode2] --> [node4] --> [node5]
+// [key3] --> [headNode3] --> [node6] --> [node7] --> [node8]
+// headNode.value is always empty, headNode is designed to make it easier to implement.
+type instancePlanCache struct {
+	heads   sync.Map
+	totCost atomic.Int64
+
+	evictMutex   sync.Mutex
+	softMemLimit atomic.Int64
+	hardMemLimit atomic.Int64
+}
+
+func (pc *instancePlanCache) getHead(key string, create bool) *instancePCNode {
+	headNode, ok := pc.heads.Load(key)
+	if ok { // cache hit
+		return headNode.(*instancePCNode)
+	}
+	if !create { // cache miss
+		return nil
+	}
+	newHeadNode := pc.createNode(nil)
+	actual, _ := pc.heads.LoadOrStore(key, newHeadNode)
+	if headNode, ok := actual.(*instancePCNode); ok { // for safety
+		return headNode
+	}
+	return nil
+}
+
+// Get gets the cached value according to key and opts.
+func (pc *instancePlanCache) Get(sctx sessionctx.Context, key string, opts any) (value any, ok bool) {
```

Comment:
ditto, It seems that it can only be PlanCacheValue.

Reply:
Yes, just to avoid cycle-import now, maybe I can change it to its concrete type later on if no cycle import problem.